server/src/handlers/bookmarks.rs

```diff
@@ -24,15 +25,38 @@ pub async fn create_bookmark(
         })
 }
 
+#[derive(Debug, Deserialize)]
+pub struct WidgetId {
+    widget_id: Id<Widget>,
+}
+
+#[axum::debug_handler]
 pub async fn get_bookmarks(
     user: User,
     State(db): State<DB>,
+    query: Option<Query<WidgetId>>,
 ) -> Result<Json<Vec<Bookmark>>, StatusCode> {
-    db.get_bookmarks(&user.id).map(Json).map_err(|e| match e {
-        DbError::NotFound => StatusCode::NOT_FOUND,
-        DbError::WhoopsieDoopsie => StatusCode::INTERNAL_SERVER_ERROR,
-        DbError::AlreadyExists => StatusCode::INTERNAL_SERVER_ERROR,
-    })
+    match query {
+        Some(w) => db
+            .get_bookmarks(&user.id)
+            .map(|v| {
+                Json(
+                    v.into_iter()
+                        .filter(|b| b.widget_id == w.widget_id)
+                        .collect::<Vec<Bookmark>>(),
+                )
+            })
+            .map_err(|e| match e {
+                DbError::NotFound => StatusCode::NOT_FOUND,
+                DbError::WhoopsieDoopsie => StatusCode::INTERNAL_SERVER_ERROR,
+                DbError::AlreadyExists => StatusCode::INTERNAL_SERVER_ERROR,
+            }),
+        None => db.get_bookmarks(&user.id).map(Json).map_err(|e| match e {
+            DbError::NotFound => StatusCode::NOT_FOUND,
+            DbError::WhoopsieDoopsie => StatusCode::INTERNAL_SERVER_ERROR,
+            DbError::AlreadyExists => StatusCode::INTERNAL_SERVER_ERROR,
+        }),
+    }
```

Comment:
I think we could reduce the duplication by moving the `match` statement *inside* the call to `.map(...)`. Because inside the `map` is the only place where the query value is potentially used.